### PR TITLE
oop-programs_with_multiple_classes-medium-q5

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MainTest extends BaseTest {
+    // Parsons
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:18]: Should be a Java question

File: src/oop/programs_with_multiple_classes/medium/q5/Desk.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package oop.programs_with_multiple_classes.medium.q5;
+
+public class Desk {
+    private String materialType;
+
+    public Desk(String m) {
+        materialType = m;
+    }
+
+    public String toString() {
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/Desk.java:10]: Remove the toString

File: src/oop/programs_with_multiple_classes/medium/q5/Monitor.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    public Monitor(int m) {
+        countSize = m;
+    }
+
+    public void fallen() {
+        countSize = countSize - 1;
+    }
+
+    public String toString() {
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/Monitor.java:14]: Remove the toString

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# oop-programs_with_multiple_classes-medium-q5
+
+**Parsons**
+
+You want to write a program that models technology and work surfaces. Make a Java program with the following requirements:
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:5]: Also, they should not be coding everything as this is a medium question

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
Please see my comments on your other PR and update accordingly

File: src/oop/programs_with_multiple_classes/medium/q5/Desk.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package oop.programs_with_multiple_classes.medium.q5;
+
+public class Desk {
+    private String materialType;
+
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/Desk.java:5]: You can also add a monitor that goes on the desk.

File: src/oop/programs_with_multiple_classes/medium/q5/Monitor.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package oop.programs_with_multiple_classes.medium.q5;
+
+public class Monitor {
+    private int countSize;
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/Monitor.java:4]: Why is this called countSize?  I feel like that does not make sense. Why don't you have `double size` attribute, and then a `isOn` boolean.

Comment (Hedgemon4) [reply src/oop/programs_with_multiple_classes/medium/q5/Monitor.java:4]: Then you can change the method in monitor to switch the value of isOn

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# oop-programs_with_multiple_classes-medium-q5
+
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:2]: Also format this file please

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+- Attributes:
+    - A `String` called **materialType**
+    - An array of `Monitor` objects called **monitors**  
+
+You must add the following items to complete the class:
+
+- Constructors:
+    - A constructor which is visible from another class and accepts a `String`
+      and initialize the **materialType** field and initializes the **monitors**
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:42]: They can initialize the array to some number I think instead

Comment (Hedgemon4) [reply src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:42]: Like an array of size x

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+## Monitor
+
+The `Monitor` class currently has the following items:
+
+- Attributes:
+    - An `int` called **size**
+    - A `boolean` called **isOn**
+
+- Constructors:
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:21]: I think they can also program this constructor

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+You want to write a program that models technology and work surfaces. Make a Java program consisting of a `Monitor`
+and a `Desk`. See the requirements below for each class. Note that parts of each class are
+already provided, you only need to implement the other portions.
+
+## Monitor
+
+The `Monitor` class currently has the following items:
+
+- Attributes:
+    - An `int` called **size**
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:18]: Please also specify the visibility modifiers

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    - A `boolean` called **isOn**
+
+- Constructors:
+    - A constructor which is visible from another class and accepts an `int` and a `boolean` as parameters
+      and initializes the **size** and **isOn** fields with the input value.
+
+You must add the following items to complete the class:
+
+- Methods:
+    - A method called **change** that changes the truth value of **isOn**;
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:28]: I would elaborate on calling this the truth value

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+The `Desk` class currently has the following items:
+
+- Attributes:
+    - A `String` called **materialType**
+    - An array of `Monitor` objects called **monitors**  
+
+You must add the following items to complete the class:
+
+- Constructors:
+    - A constructor which is visible from another class and accepts a `String`
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:41]: You need to specify the visibility of the constructor

File: src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md (hunk 11 above)
Comment (Hedgemon4) [reply src/oop/programs_with_multiple_classes/medium/q5/QuestionText.md:28]: Also the method needs an accessibility modifier and return type

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("monitorInputProvider")
+    public void correctFallenMethod(int count, boolean on) throws Throwable {
+        Object[][] arguments = {
+                {count, int.class},
+                {on, boolean.class}
+        };
+        Object monitorInstance = monitor.createInstance(arguments);
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:116]: Please also test the return type of the method separately. You can do this with `hasMethod`, and specify the return type to be `Void.TYPE`

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("deskInputProvider")
+    public void deskConstructorInitializesValuesCorrectly(String material) throws Throwable {
+        Object[][] arguments = {
+                {material, String.class},
+        };
+        Object chairInstance = desk.createInstance(arguments);
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:94]: You should also check the visibility modifier of the constructor

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java (hunk 14 above)
Comment (KaiKhademi) [reply src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:94]: We check that in deskClassHasRequiredConstructor()

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    @ParameterizedTest
+    @MethodSource("monitorInputProvider")
+    public void correctFallenMethod(int count, boolean on) throws Throwable {
+        Object[][] arguments = {
+                {count, int.class},
+                {on, boolean.class}
+        };
+        Object monitorInstance = monitor.createInstance(arguments);
+        monitor.callMethod(fallenMethod, monitorInstance);
+        monitor.hasMethod(fallenMethod, new Class[]{int.class, boolean.class}, Void.TYPE);
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:137]: This needs to be in an assertTrue as it returns a boolean

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+    @ParameterizedTest
+    @MethodSource("monitorInputProvider")
+    public void correctFallenMethod(int count, boolean on) throws Throwable {
+        Object[][] arguments = {
+                {count, int.class},
+                {on, boolean.class}
+        };
+        Object monitorInstance = monitor.createInstance(arguments);
+        assertTrue(monitor.hasMethod(fallenMethod, null, Void.TYPE),
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:136]: Please separate these out into different asserts. You should check the method header, and return type separately

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+public class MainTest {
+    // Java
+
+    private final String deskClass = "Desk";
+    private final String monitorClass = "Monitor";
+    private final String varMaterial = "materialType";
+    private final String varMonitors = "monitors";
+    private final String varCount = "size";
+    private final String varOn = "isOn";
+    private final String fallenMethod = "change";
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:25]: Can we change this to not be called fallenMethod

File: src/oop/programs_with_multiple_classes/medium/q5/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    @ParameterizedTest
+    @MethodSource("monitorInputProvider")
+    public void correctFallenMethod(int count, boolean on) throws Throwable {
+        Object[][] arguments = {
+                {count, int.class},
+                {on, boolean.class}
+        };
+        Object monitorInstance = monitor.createInstance(arguments);
+        assertTrue(monitor.hasMethod(fallenMethod, null, Void.TYPE),
+                "Your " + monitorClass + " does not include the correct " + fallenMethod);
```
Comment (Hedgemon4) [inline src/oop/programs_with_multiple_classes/medium/q5/MainTest.java:137]: Also please check the visibility modifier

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**